core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -1548,6 +1591,327 @@ public void addStateChangeListener(StateChangeListener<DistributedStagesSchedule
+                NodeAllocator nodeAllocator,
+                ScheduledFuture<?> nodeUpdateTask)
+        {
+            this.stateMachine = requireNonNull(stateMachine, "stateMachine is null");
+            this.queryStateMachine = requireNonNull(queryStateMachine, "queryStateMachine is null");
+            this.schedulers = requireNonNull(schedulers, "schedulers is null");
+            this.schedulerStats = requireNonNull(schedulerStats, "schedulerStats is null");
+            this.nodeAllocator = requireNonNull(nodeAllocator, "nodeAllocator is null");
+            this.nodeUpdateTask = requireNonNull(nodeUpdateTask, "nodeUpdateTask is null");
+        }
+
+        @Override
+        public void schedule()
+        {
+            checkState(started.compareAndSet(false, true), "already started");
+
+            try (SetThreadName ignored = new SetThreadName("Query-%s", queryStateMachine.getQueryId())) {
+                List<ListenableFuture<Void>> blockedStages = new ArrayList<>();
+                while (!isFinishingOrDone(queryStateMachine) && !stateMachine.getState().isDone()) {
+                    blockedStages.clear();
+                    boolean atLeastOneStageIsNotBlocked = false;
+                    boolean allFinished = true;
+                    for (FaultTolerantStageScheduler scheduler : schedulers) {
+                        if (scheduler.isFinished()) {
+                            continue;
+                        }
+                        allFinished = false;
+                        ListenableFuture<Void> blocked = scheduler.isBlocked();
+                        if (!blocked.isDone()) {
+                            blockedStages.add(blocked);
+                            continue;
+                        }
+                        try {
+                            scheduler.schedule();
+                        }
+                        catch (Throwable t) {
+                            stateMachine.transitionToFailed(t, Optional.of(scheduler.getStageId()));
+                            return;
+                        }
+                        blocked = scheduler.isBlocked();
+                        if (!blocked.isDone()) {
+                            blockedStages.add(blocked);
+                        }
+                        else {
+                            atLeastOneStageIsNotBlocked = true;
+                        }
+                    }
+                    if (allFinished) {
+                        stateMachine.transitionToFinished();
+                        return;
+                    }
+                    // wait for a state change and then schedule again
+                    if (!atLeastOneStageIsNotBlocked && !blockedStages.isEmpty()) {
+                        try (TimeStat.BlockTimer timer = schedulerStats.getSleepTime().time()) {
+                            try {
+                                tryGetFutureValue(whenAnyComplete(blockedStages), 1, SECONDS);
```

Comment:
Cant you attach listener to `ListenableFuture` returned by `whenAnyComplete`?

Reply:
We can. I was thinking about it. But if there's any subtle bug in scheduler (when something doesn't get unlocked when needed) there's a chance the query might get stuck. So I thought it might be better to release once in a while. The same technique is used for Pipelined scheduling.